src/store.cc

```diff
@@ -2074,10 +2075,12 @@ StoreEntry::describeTimestamps() const
 bool
 StoreEntry::collapsingInitiator() const
 {
-    if (!publicKey())
-        return false;
-    return EBIT_TEST(flags, ENTRY_FWD_HDR_WAIT) ||
-           (hasTransients() && !hasMemStore() && !hasDisk());
+    // TODO: the method description says that this must be
+    // a collapsed forwarding-created public entry, however
+    // this implementation has no explicit confirmation of that.
+    // We should verify here that collapsing has been allowed
+    // for this entry by Store::Controller::allowCollapsing().
+    return publicKey() && isEmpty();
```

Comment:
Both isEmpty() and `ENTRY_FWD_HDR_WAIT` are dirty hacks -- when they work, they work essentially by accident. You have already discovered cases where `ENTRY_FWD_HDR_WAIT` does not work. And a method based on raw stored bytes, like isEmpty(), may fail if we start storing partial headers or perhaps some other entry metadata _before_ its cachability status is known.

As you probably know by now, there is more about this problematic method, including a better name/description [elsewhere](https://github.com/measurement-factory/squid/commit/574b37469b5f604690bd94418bd904bf25b2465f).

Perhaps we should bite the bullet and add an `ENTRY_REQUIRES_COLLAPSING` flag instead of failing to find a reliable way to identify these entries using existing metadata?

Reply:
I think we have to use this new ENTRY_REQUIRES_COLLAPSING (addressing this TODO), however, we still should use isEmpty(), because this method must distinguish collapsed hits from regular hits. So, it would be:
```
bool
StoreEntry::collapsingInitiator() const
{
    return publicKey() && isEmpty() &&  EBIT_TEST(flags, ENTRY_REQUIRES_COLLAPSING);
}
```
Do you agree?

Reply:
> we do not create a new 'accidental' case, but make use of an already existing principle: the entry is empty until reply headers are parsed.

startWriting() _naturally_ asserts that nobody has written anything yet when calling a method called "start writing". One can adjust code to save partial headers (or some metadata) to disk earlier. The existing isEmpty() assertion in startWriting() will continue to work just fine after that adjustment (but our code will break). There is no natural connection or cause/effect relationship between the entry having no header/body bytes in RAM and the entry response having unknown/not-yet-determined cachability status. Relying on that non-existent relationship is a hack.

I am especially worried that isEmpty() is about MemObject state. What if I store headers (and even a whole or a part of the body) and then completely purge all of that data from local memory? Would not isEmpty() be true, even though the entry cachability is already known?

The big question is: Is adding (and updating) a new StoreEntry flag better than relying on an unnatural "isEmpty() == cachability unknown" dependency?

Reply:
> The big question is

I think another big(conceptual) question is to understand when collapsing can happen and when already cannot. Obviously, it can happen just after a public entry is created (Controller::allowCollapsing() returned true). But after which event we can say that the collapsing is already impossible? Earlier I thought that it is already impossible just after the entry has got the first data byte (isEmpty() is false). It seems now that this assumption is wrong.

Reply:
I do not know the answer to that important question, but I would look at the entry synchronization code: Perhaps the flag can be cleared when the StoreEntry object is attached to one of the cache stores? After that point, there are no CF-specific risks associated with hitting the entry (compared to regular cache hits for incomplete entries).

Reply:
Done at 6a0dd41 (merged d2a6dcb (PR213) from master).